critical_section.c

```diff
@@ -0,0 +1,24 @@
+#include <critical_section.h>
+#include <interrupts.h>
+#include <common.h>
+#include <libkern.h>
+
+volatile int cs_level = 0;
+
+void cs_enter() {
+  if (during_intr_handler())
+    return;
+  intr_disable();
+  cs_level++;
+}
+
+void cs_leave() {
+  if (during_intr_handler())
+    return;
+
+  assert(cs_level > 0);
+
+  cs_level--;
+  if (cs_level == 0)
+    intr_enable();
```

Comment:
This will break if someone enters critical section with interrupts disabled by direct call to `intr_disable`. General practice is to have interrupt disabling function to return the previous state so that it can be restored when the section requiring to be not interrupted is left.